overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+		}
+    }
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+	/// The actual async task running.
+	f: SubsystemJob,
+}
+
+/// An `id` that is given to any `SubsystemInstance` for identification.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct SubsystemId(usize);
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M>>,
+}
+
+impl<M: Debug> SubsystemContext<M> {
+	/// Try to asyncronously receive a message.
+	///
+	/// This has to be used with caution, if you loop over this without
+	/// using `pending!()` macro you will end up with a busy loop!
+	pub async fn try_recv(&mut self) -> Result<Option<M>, ()> {
+		match poll!(self.rx.next()) {
+		    Poll::Ready(Some(msg)) => Ok(Some(msg)), 
+			Poll::Ready(None) => Err(()),
+		    Poll::Pending => Ok(None),
+		}
+	}
+
+	/// Receive a message.
+	pub async fn recv(&mut self) -> Result<M, SubsystemError> {
+		self.rx.next().await.ok_or(SubsystemError)
+	}
+
+	/// Send a message to whom it may concern.
+	///
+	/// The message will be broadcasted to all other `Subsystem`s that can
+	/// receive it.
+	pub async fn send_msg(&mut self, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
+			to: None,
+			msg,
+		}).await;
+	}
+
+	/// Spawn a child `Subsystem` on the executor and get it's `SubsystemId` upon success.
+	pub async fn spawn(&mut self, s: Box<dyn Subsystem<M>>) -> Result<SubsystemId, SubsystemError> {
+		let (tx, rx) = oneshot::channel();
+		let _ = self.tx.send(OverseerMessage::Spawn{
+			s,
+			res: tx,
+		}).await;
+
+		rx.await.unwrap_or_else(|_| Err(SubsystemError))
+	}
+
+	/// Send a direct message to some other `Subsystem` you know `SubsystemId` of.
+	pub async fn send_msg_to(&mut self, to: SubsystemId, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
+			to: Some(to),
+			msg,
+		}).await;
+	}
+
+	fn new(rx: mpsc::Receiver<M>, tx: mpsc::Sender<OverseerMessage<M>>) -> Self {
+		Self {
+			rx,
+			tx,
+		}
+	}
+}
+
+/// A trait that describes the `Subsystems` that can run on the `Overseer`.
+///
+/// It is generic over the message type circulating in the system.
+/// The idea that we want some type contaning persistent state that
+/// can start actually running jobs when asked to.
+pub trait Subsystem<M: Debug> {
+	/// Start this `Subsystem` and return `SubsystemJob`.
+	fn start(&mut self, ctx: SubsystemContext<M>) -> SubsystemJob;
+	/// If this `Subsystem` want to receive this message.
+	///
+	/// By default receive all messages.
+	fn can_recv_msg(&self, _msg: &M) -> bool { true }
+}
```

Comment:
I'm not convinced that all subsystems will fit a trait very well. I'd like the subsystems to fit a certain pattern, but I don't want them to be constrained by a trait like this.

I'm more keen on the 'big struct with a load of fields' design that @rphmeier talked about here: https://github.com/w3f/research/pull/84#issuecomment-624424121

Reply:
This seems hard to design if you don't know what the subsystems are going to be like first though.

Reply:
I like it, it's generic over `M` and decoupled via `channel`s - matches up with what I had in mind :+1:

Reply:
@drahnr no, the `M` is one-message-type-to-rule-them-all in the system, something like

```rust
enum Message {
 NetworkMessage,
 StorageMessage,
 ValidationMessage,
 ..
}
```

Reply:
Under what circumstances would a subsystem not want to receive a message?

Reply:
Well, actually in that proposal subsystem A receives `FromOverseer<AMessage>`

```rust
enum FromOverseer<T> {
    Signal(OverseerSignal),
    Communication(T),
}
```